[SPARK-13526] [SQL] Move SQLContext per-session states to new class

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

This creates a `SessionState`, which groups a few fields that existed in `SQLContext`. Because `HiveContext` extends `SQLContext` we also need to make changes there. This is mainly a cleanup task that will soon pave the way for merging the two contexts.
## How was this patch tested?

Existing unit tests; this patch introduces no change in behavior.

Commit:
Add SessionState template

Commit:
Move fields to SessionState for SQL/HiveContext

Commit:
Clean up code + add/restore comments

andrewor14 commented:
@rxin

SparkQA commented:
**[Test build #52092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52092/consoleFull)** for PR 11405 at commit [`814ae1d`](https://github.com/apache/spark/commit/814ae1d5ccbb6680e808b280de9c454f76113f28).

SparkQA commented:
**[Test build #52092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52092/consoleFull)** for PR 11405 at commit [`814ae1d`](https://github.com/apache/spark/commit/814ae1d5ccbb6680e808b280de9c454f76113f28).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52092/
Test FAILed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SessionState.scala:
```diff
@@ -0,0 +1,108 @@
+/**
+ * A class that holds all session-specific state in a given [[SQLContext]].
+ */
+private[sql] class SessionState(ctx: SQLContext) {
```
put this in the new internal package?

andrewor14 replied:
which one?

rxin replied:
o.a.s.sql.internal

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:
```diff
@@ -34,9 +35,7 @@ import org.apache.spark.sql.types.DataType
-class UDFRegistration private[sql] (sqlContext: SQLContext) extends Logging {
-
-  private val functionRegistry = sqlContext.functionRegistry
+class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends Logging {
```
this is a great change!

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -110,6 +108,12 @@ class HiveContext private[hive](
+
+  protected[sql] override val sessionState = hiveSessionState
+
+  protected[hive] val hcatalog = hiveSessionState.metastoreCatalog
```
hcatalog is a bad name because there is actually a hive-related project called hcatalog. Maybe just call this hiveCatalog?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -115,9 +113,26 @@ class SQLContext private[sql](
-  protected[sql] lazy val conf = new SQLConf
+  @transient
+  protected[sql] val sessionState: SessionState = new SessionState(self)
+  protected[sql] def conf: SQLConf = sessionState.conf
```
we will probably need to remove these once we create SparkSession and pass that around - but for now it is ok to leave them here.

rxin commented:
Looks pretty good overall.

Commit:
Fix MiMa

Commit:
Move SessionState to internal package

Commit:
hcatalog -> hiveCatalog

SparkQA commented:
**[Test build #52093 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52093/consoleFull)** for PR 11405 at commit [`fe0f792`](https://github.com/apache/spark/commit/fe0f792c5483887cbc777bd518a93c99cb33d2dd).

Commit:
Merge branch 'master' of github.com:apache/spark into refactor-session

Conflicts:
	project/MimaExcludes.scala

Commit:
Fix style

SparkQA commented:
**[Test build #52093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52093/consoleFull)** for PR 11405 at commit [`fe0f792`](https://github.com/apache/spark/commit/fe0f792c5483887cbc777bd518a93c99cb33d2dd).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52093/
Test FAILed.

andrewor14 commented:
retest this please

andrewor14 commented:
OK, I think I addressed all of your comments.

SparkQA commented:
**[Test build #52094 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52094/consoleFull)** for PR 11405 at commit [`60852ff`](https://github.com/apache/spark/commit/60852fffd04839d75069865fb9e5d4b5a0549f32).

rxin commented:
LGTM

SparkQA commented:
**[Test build #52094 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52094/consoleFull)** for PR 11405 at commit [`60852ff`](https://github.com/apache/spark/commit/60852fffd04839d75069865fb9e5d4b5a0549f32).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52094/
Test FAILed.

rxin commented:
Compilation failed.

Commit:
Fix test compile

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #52104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52104/consoleFull)** for PR 11405 at commit [`775f6c2`](https://github.com/apache/spark/commit/775f6c2d13fc388a24a4f0f3d1ee05b2e7fcac5c).

Commit:
Fix test compile for reals

SparkQA commented:
**[Test build #52106 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52106/consoleFull)** for PR 11405 at commit [`c9af485`](https://github.com/apache/spark/commit/c9af4857807e0d0219a87292d72f8bc7a0f59f89).

rxin commented:
:)

SparkQA commented:
**[Test build #52106 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52106/consoleFull)** for PR 11405 at commit [`c9af485`](https://github.com/apache/spark/commit/c9af4857807e0d0219a87292d72f8bc7a0f59f89).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52106/
Test FAILed.

andrewor14 commented:
Tests are failing locally. I'm investigating.

Commit:
Fix style

SparkQA commented:
**[Test build #52104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52104/consoleFull)** for PR 11405 at commit [`775f6c2`](https://github.com/apache/spark/commit/775f6c2d13fc388a24a4f0f3d1ee05b2e7fcac5c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52104/
Test FAILed.

Commit:
Fix tests

Stupid constructor initialization ordering.

Commit:
Make diff smaller + fix Hive tests

It turns out when you override a val and you don't specify the
type, you can make it of a different type (one that subclasses
the parent type). This makes it possible to revert some of the
previous changes related to the `hiveCatalog` field.

SparkQA commented:
**[Test build #52113 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52113/consoleFull)** for PR 11405 at commit [`ee3d95b`](https://github.com/apache/spark/commit/ee3d95bd3117bb48cc2a833991866b9b08dde483).

andrewor14 commented:
retest this please

andrewor14 commented:
OK, latest round should pass tests.

SparkQA commented:
**[Test build #52114 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52114/consoleFull)** for PR 11405 at commit [`d151fd1`](https://github.com/apache/spark/commit/d151fd107e42697de78e251b11043f8e8e7d8cc2).

SparkQA commented:
**[Test build #52113 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52113/consoleFull)** for PR 11405 at commit [`ee3d95b`](https://github.com/apache/spark/commit/ee3d95bd3117bb48cc2a833991866b9b08dde483).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52113/
Test FAILed.

SparkQA commented:
**[Test build #52114 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52114/consoleFull)** for PR 11405 at commit [`d151fd1`](https://github.com/apache/spark/commit/d151fd107e42697de78e251b11043f8e8e7d8cc2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52114/
Test FAILed.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -115,9 +113,27 @@ class SQLContext private[sql](
-  protected[sql] lazy val conf = new SQLConf
+  @transient
+  protected[sql] lazy val sessionState: SessionState = new SessionState(self)
+  protected[sql] lazy val conf: SQLConf = sessionState.conf
```
these lazy vals should probably be defs?

andrewor14 replied:
they could be, but the underlying things in session state aren't gonna change. Should I make them defs?

rxin replied:
Yea def is better - just reducing the number of fields and you also don't need to make them transients.

Commit:
Make lazy vals defs

Commit:
Make sure sessionState is transient in subclasses

Commit:
Fix tests

Hive current_database needs to be registered in the constructor
of the HiveContext. Before this commit, that was not the case
because HiveSessionState's functionRegistry is lazy. The original
behavior before this patch is now restored as of this commit.

andrewor14 commented:
Alright, all previously failing tests pass locally. Jenkins retest this please.

SparkQA commented:
**[Test build #52130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52130/consoleFull)** for PR 11405 at commit [`c76a095`](https://github.com/apache/spark/commit/c76a095de92392f53a1bedd8a4bf57a1ca86b6b7).

SparkQA commented:
**[Test build #52130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52130/consoleFull)** for PR 11405 at commit [`c76a095`](https://github.com/apache/spark/commit/c76a095de92392f53a1bedd8a4bf57a1ca86b6b7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52130/
Test FAILed.

Commit:
Fix tests better

SparkQA commented:
**[Test build #52132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52132/consoleFull)** for PR 11405 at commit [`bdf1a5d`](https://github.com/apache/spark/commit/bdf1a5d5a0cc60b6fc8cf5a6c01d75ad6c5b12aa).

SparkQA commented:
**[Test build #52132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52132/consoleFull)** for PR 11405 at commit [`bdf1a5d`](https://github.com/apache/spark/commit/bdf1a5d5a0cc60b6fc8cf5a6c01d75ad6c5b12aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52132/
Test PASSed.

rxin commented:
Merging this in master.
